src/components/data-pane/field-list.tsx

```diff
@@ -58,6 +80,40 @@ class FieldListBase extends React.PureComponent<FieldListProps, {}> {
       payload: {fieldDef: fieldDef}
     });
   }
+
+  private renderTypeChanger(fieldDef: ShelfFieldDef, primitiveType: PrimitiveType) {
+    const {handleAction} = this.props;
+    if (typeof fieldDef.field === 'string' && this.state.selectedField === fieldDef.field) {
+      return (
+        <TypeChanger
+          field={fieldDef.field}
+          type={fieldDef.type}
+          types={this.getTypes(primitiveType)}
+          handleAction={handleAction}
+        />
+      );
+    }
+  }
+
+  private handleCaretClick(field: string) {
+    if (this.state.selectedField === field) {
+      this.setState({
+        selectedField: null
+      });
+    } else {
+      this.setState({
+        selectedField: field
+      });
+    }
+  }
+
+  private getTypes(primitiveType: PrimitiveType): ExpandedType[] {
+    if (primitiveType === PrimitiveType.NUMBER) {
```

Comment:
Let's do a `switch` and correctly determine types for all primitive types?

(string = NOMINAL only, datetime = TEMPORAL only, etc.)